students/cjfortu/L07/assignment/database_multithread_join.py

```diff
@@ -0,0 +1,310 @@
+
+
+def print_database():
+    """
+    Present the full database.
+    """
+    mongo = MongoDBConnection()
+    with mongo:
+        db = mongo.connection['products_database']
+        products_collection = db["products"]
+        customers_collection = db["customers"]
+        rentals_collection = db["rentals"]
+
+    collections = [products_collection, customers_collection, rentals_collection]
+    for collection in collections:
+        print(collection)
+        for document in collection.find():
+            document.pop('_id', None)
+            print(document)
+
+
+def import_data(directory_name, product_file, customer_file, rentals_file):
+    """
+    Populate the new MongoDB database with source data, and return two tuples:
+
+    1: A record count of the number of products, customers, and rentals added (in this order)
+    2: A count of any errors that occured, in the same order.
+    """
+    file_paths = [directory_name + product_file,
+                  directory_name + customer_file,
+                  directory_name + rentals_file]
+    error_counts = [0, 0, 0]
+    document_counts = [0, 0, 0]
+
+    init = time.process_time()
+
+    mongo = MongoDBConnection()
+    with mongo:
+        db = mongo.connection['products_database']
+        products_collection = db["products"]
+        customers_collection = db["customers"]
+        rentals_collection = db["rentals"]
+
+    collections = [products_collection, customers_collection, rentals_collection]
+
+    def write_data(file_path, collection, i):
+        """
+        Helper function to provide a target for multithreaded writing.
+        """
+        try:
+            with open(file_path, 'r', encoding='utf-8-sig') as csv_file:
+                csv_reader = csv.reader(csv_file)
+                fields = next(csv_reader)
+                extracted_data = []
+                for row in csv_reader:
+                    row_dict = {}
+                    for j, field in enumerate(fields):
+                        error = False
+                        if row[j].strip() == '' and field != 'end_date':
+                            error = True
+                            break
+                        row_dict[field] = row[j]
+                    if error is False:
+                        extracted_data.append(row_dict)
+                    else:
+                        error_counts[i] += 1
+        except FileNotFoundError as err:
+            bad_file_path = True
+            print(err)
+            LOGGER.info(f'{collection} not added due to FileNotFoundError.')
+        else:
+            collection.insert_many(extracted_data)
+            LOGGER.info(f'{collection} successfully added.')
+            document_counts[i] = collection.count_documents({})
+
+    bad_file_path = False
+    for i in range(len(file_paths)):
+        thread = threading.Thread(target=write_data, args=(file_paths[i], collections[i], i))
+        thread.start()
+        thread.join()
```

Comment:
If you run your _join()_ immediately after launching a thread or multiprocess, you essentially force it to be sequential. This is because _join()_ will sit there and wait for the thread to finish. Instead, do something like this:
   
```
    thread_list = []
    for i in range(len(file_paths)):
        thread = threading.Thread(target=write_data, args=(file_paths[i], collections[i], i))
        thread.start()
        thread_list.append(thread)

    for thread in thread_list():
        thread.join()
```

This will launch all threads, let them run and then wait for each one to finish. You will need to use queues for each thread to report back and consider that fact that threads might put the results in the queue in a different order from the order in which they were launched.

This will also work for multiprocessing, consider that you need to use _multiprocessing.queue_, other than that, the implementation should be identical to threading.

I'm wondering why your multiprocessing was so much faster, even with you joining each process right away. My guess is that a dedicated core still executes faster than one that is already dealing with other operations.